src/proxy/ssh/AgentProxy.ts

```diff
@@ -0,0 +1,306 @@
+      }
+      const publicKeyBlob = response.slice(offset, offset + blobLength);
+      offset += blobLength;
+
+      // Read comment length
+      if (response.length < offset + 4) {
+        throw new Error(`Invalid identities response: missing comment length for key ${i}`);
+      }
+      const commentLength = response.readUInt32BE(offset);
+      offset += 4;
+
+      // Read comment
+      if (response.length < offset + commentLength) {
+        throw new Error(`Invalid identities response: incomplete comment for key ${i}`);
+      }
+      const comment = response.slice(offset, offset + commentLength).toString('utf8');
+      offset += commentLength;
+
+      // Extract algorithm from key blob (SSH wire format: [length:4][algorithm string])
+      let algorithm = 'unknown';
+      if (publicKeyBlob.length >= 4) {
+        const algoLen = publicKeyBlob.readUInt32BE(0);
+        if (publicKeyBlob.length >= 4 + algoLen) {
+          algorithm = publicKeyBlob.slice(4, 4 + algoLen).toString('utf8');
+        }
+      }
+
+      identities.push({ publicKeyBlob, comment, algorithm });
+    }
+
+    return identities;
+  }
+
+  /**
+   * Request the agent to sign data with a specific key
+   *
+   * @param publicKeyBlob - The public key blob identifying which key to use
+   * @param data - The data to sign
+   * @param flags - Signing flags (usually 0)
+   * @returns The signature blob
+   */
+  async sign(publicKeyBlob: Buffer, data: Buffer, flags: number = 0): Promise<Buffer> {
+    // Build SIGN_REQUEST message
+    // Format: [type:1][key_blob_len:4][key_blob][data_len:4][data][flags:4]
+    const message = Buffer.concat([
+      Buffer.from([AgentMessageType.SSH_AGENTC_SIGN_REQUEST]),
+      this.encodeBuffer(publicKeyBlob),
+      this.encodeBuffer(data),
+      this.encodeUInt32(flags),
+    ]);
+
+    const response = await this.sendMessage(message);
+
+    // Parse response
+    const responseType = response[0];
+
+    if (responseType === AgentMessageType.SSH_AGENT_FAILURE) {
+      throw new Error('Agent returned failure for sign request');
+    }
+
+    if (responseType !== AgentMessageType.SSH_AGENT_SIGN_RESPONSE) {
+      throw new Error(`Unexpected response type: ${responseType}`);
+    }
+
+    // Parse signature
+    // Format: [type:1][sig_blob_len:4][sig_blob]
+    if (response.length < 5) {
+      throw new Error('Invalid sign response: too short');
+    }
+
+    const sigLength = response.readUInt32BE(1);
+    if (response.length < 5 + sigLength) {
+      throw new Error('Invalid sign response: incomplete signature');
+    }
+
+    const signatureBlob = response.slice(5, 5 + sigLength);
+
+    // The signature blob format from the agent is: [algo_len:4][algo:string][sig_len:4][sig:bytes]
+    // But ssh2 expects only the raw signature bytes (without the algorithm wrapper)
+    // because Protocol.authPK will add the algorithm wrapper itself
+
+    // Parse the blob to extract just the signature bytes
```

Comment:
Are we 100% sure that the SSH signature format will always be the same? Just wondering if support for alternative algorithms is necessary.

If SSH connections to GH always use the same algorithm, maybe this isn't needed, but I'm not sure if connections between the user and GitProxy might require additional support.

Reply:
A user could use any key type that GitHub or another provider supports, so for example U2F keys use [sk-ecdsa-sha2-nistp256@openssh.com](https://github.com/openssh/openssh-portable/blob/8fce5520a1c9c2cf3fc6c6974dd158f4b3ce9c4e/PROTOCOL.u2f#L204), which as you can see includes a flags and counter additionally to the signature.

My feeling would be to get this PR in with the common key types and maybe add a todo/issue for some further testing.